packages/compartment-mapper/src/policy.js

```diff
@@ -0,0 +1,304 @@
+// @ts-check
+
+/** @typedef {import('./types.js').PolicyIdentityComponents} PolicyIdentityComponents */
+
+const { create, entries, assign, keys, freeze } = Object;
+const { stringify } = JSON;
+
+const copyGlobals = (globals, list) => {
+  const g = create(null);
+  if (list && list.length > 0) {
+    for (let k = 0; k < list.length; k += 1) {
+      const key = list[k];
+      // If an endowment is missing, global value is undeinfed.
+      // This is an expected behavior if globals are used for platform feature detection
+      g[key] = globals[key];
+    }
+  }
+  return g;
+};
+
+/**
+ * Const string to identify the internal attenuators compartment
+ */
+export const ATTENUATORS_COMPARTMENT = '<ATTENUATORS>';
+
+const attenuatorsCache = new WeakMap();
+/**
+ * Goes through policy and lists all attenuator specifiers used.
+ * Memoization keyed on policy object reference
+ *
+ * @param {*} policy
+ * @returns {Array<string>} attenuators
+ */
+export const detectAttenuators = policy => {
+  if (!attenuatorsCache.has(policy)) {
+    const attenuators = [];
+    // a free recursive visitor implementation for simple objects
+    stringify(policy, (key, value) => {
+      if (key === 'attenuate' && typeof value === 'string') {
+        attenuators.push(value);
+      }
+      return value;
+    });
+    attenuatorsCache.set(policy, attenuators);
+  }
+  return attenuatorsCache.get(policy);
+};
+
+/**
+ * Generates an ID for the package for policy purposes.
+ * Some arguments here are for future proofing - I could imagine using them for IDs instead
+ *
+ * @param {PolicyIdentityComponents} identityComponents
+ * @returns {string}
+ */
+const generatePolicyId = ({ isEntry = false, name, path }) => {
```

Comment:
I'm considering adding location here as bikeshedding for hupothetical future use